File: Csharp/RT_CS/RT_CS/shapes/Triangle.cs
using RT_CS.Cameras;
using RT_CS.utility;
using System;
using System.Collections.Generic;
using System.Text;

namespace RT_CS.shapes
{

    public class Triangle : Shape
    {
        protected Vector3 vertexA, vertexB, vertexC, normal;
        protected Vector3 D;
        private Vector3 ab,ac,bc;
        protected float area;
        private float uTop;

        protected const float BIGGER_EPSILON = 0.000001f;


        protected static float SimplerTriangleArea(Vector3 ab,Vector3 ac)
        {
            return (ab * ac).Magnitude() * 0.5f;
        }

        public Triangle(int id, int mid, Vector3 v1, Vector3 v2, Vector3 v3)  :base(id, mid)
        {
            vertexA = v1;
            vertexB = v2;
            vertexC = v3;
            ab = vertexB - vertexA;
            ac = vertexC - vertexA;
            bc = vertexC - vertexA;
            normal = ((v3 - v1) * (v2 - v1)).Normalize();
            D = v1;
            area = SimplerTriangleArea(ab,ac);
            float acMag = ac.Magnitude();
            uTop = Vector3.Dot(ab, ac / acMag) / acMag;
        }
        public override Vector3 GetNormal(Vector3 point)
        {
            return normal;
        }
        public IntersectionRecord RayPlaneIntersection(Ray r)
        {
            IntersectionRecord inter = new IntersectionRecord();
            float nordir = Vector3.Dot(r.direct, normal);
            if (Math.Abs(nordir) < float.Epsilon)
            {
                return inter;
            }
            else
            {
                float d = Vector3.Dot((D - r.posit),normal)/nordir;
                if (d > 0)
                {
                    inter.Distance = d;
                    inter.Hit = true;
                    inter.Point = r.posit + d * r.direct;
                }
                return inter;
            }
        }

        public override Vector2 GetUV(Vector3 pt)
        {
            float areaAlpha = SimplerTriangleArea(vertexB - vertexA, pt - vertexA) / area;
            float areaBeta = SimplerTriangleArea(pt - vertexB, pt - vertexC) / area;
            float areaGamma = SimplerTriangleArea(pt - vertexA, vertexC - vertexA) / area;

            return new Vector2 { x = (uTop * areaBeta) + areaGamma, y = areaAlpha + areaGamma };
        }

        public override void Intersect(Ray r, IntersectionRecord intersection)
        {
            if (intersection.ExceptId == sid)
            {
                return;
            }
            IntersectionRecord inter = RayPlaneIntersection(r);
            if (inter.Hit)
            {
                float areaAlpha = SimplerTriangleArea(vertexB - vertexA, inter.Point - vertexA) / area;
                float areaBeta = SimplerTriangleArea(inter.Point - vertexB, inter.Point - vertexC) / area;
                float areaGamma = SimplerTriangleArea(inter.Point - vertexA, vertexC - vertexA) / area;
                float sum = areaAlpha + areaBeta + areaGamma;


                if(Math.Abs(1-sum) < BIGGER_EPSILON)
                {
                    if (inter.Distance < intersection.Distance && inter.Distance > 0)
                    {
                        intersection.Distance = inter.Distance;
                        intersection.Hit = true;
                        intersection.Point = inter.Point;
                        intersection.ShapeId = sid;
                    }
                }
            }
        }
    }


}

File: cpp/RT_CPP/RT_CPP/triangle.cpp
#include "triangle.h"
#include "operators.h"
#include <math.h>
#include <cfloat>
using namespace shapes;

triangle::triangle(int id,int mid, vector3 a, vector3 b, vector3 c) : shape(id,mid),
vertexA(a), vertexB(b), vertexC(c)
{
	ab = vertexB - vertexA;
	ac = vertexC - vertexA;
	bc = vertexC - vertexB;
	normal = (ac * ab).normalize();
	area = simpler_triangle_area(ab, ac);
	d = vertexA;
	float acMag = ac.magnitude();
	u_top = vector3::dot(ab, ac / acMag) / acMag;
}

float triangle::simpler_triangle_area(vector3 ab, vector3 ac)
{
	return 0.5f * (ab * ac).magnitude();
}

intersection_record* triangle::ray_plane_intersection(ray r) const
{
	intersection_record* rec = new intersection_record;
	
	float dot = vector3::dot(r.direction, normal);
	if (fabsf(dot) < FLT_EPSILON)
	{
		return rec; // parallel to the plane, no intersection
	}
	float dist = (vector3::dot(normal,d - r.position)) / dot;
	
	if (dist > 0)
	{
		rec->distance = dist;
		rec->hit = true;
		rec->point = r.position + dist * r.direction;
	}

	return rec;
	
}

void triangle::intersect(ray r, intersection_record& rec) const
{
	if (id == rec.except_id)
	{
		return;
	}
	intersection_record* ray_int = ray_plane_intersection(r);
	if (ray_int->hit)
	{
		//computing barymetric area ratios
		float areaAlpha = simpler_triangle_area(vertexB - vertexA, ray_int->point - vertexA) /area;
		float areaBeta = simpler_triangle_area(ray_int->point - vertexB, ray_int->point - vertexC) / area;
		float areaGamma = simpler_triangle_area(ray_int->point - vertexA, vertexC - vertexA) /area;

		float sum = areaAlpha + areaBeta + areaGamma;
		if (fabsf(1 - sum) < BIGGER_EPSILON)
		{
			if (ray_int->distance < rec.distance && ray_int->distance > 0)
			{
				rec.hit = true;
				rec.distance = ray_int->distance;
				rec.shape_id = id;
				rec.point = ray_int->point;
			}
		}
	}
	delete ray_int;
}

vector2 triangle::get_uv(vector3 point) const
{
	//computing barymetric area ratios
	float areaAlpha = simpler_triangle_area(vertexB - vertexA, point - vertexA) / area;
	float areaBeta = simpler_triangle_area(point - vertexB, point - vertexC) / area;
	float areaGamma = simpler_triangle_area(point - vertexA, vertexC - vertexA) / area;

	float u = u_top * areaBeta + areaGamma;
	float v = areaAlpha + areaGamma;

	return { u,v };
}

vector3 triangle::get_normal(vector3 point) const
{
	return normal;
}